docs/design/rosbag2_snapshot_recorder/rosbag2_snapshot_recorder.md

```diff
@@ -0,0 +1,738 @@
+# Rosbag2 Snapshot
+
+## Context
+
+The following design handles functionality of keeping data from topics in buffer of specified size or from specific time duration and saving them after trigger.
+
+Terms:
+* "Snapshot" - save buffered data from topics to disk as rosbag
+* "Trigger" - an invocation letting a node know when to save rosbag to disk
+* "Recorder" - rosbag2 node which records data right after program start until killed
+
+## Feature description
+
+Rosbag in ROS1 had functionality of creating a [snapshot](https://github.com/ros/rosbag_snapshot). Snapshotter was a program which subscribed to topics and buffered messages until a `trigger` for snapshot was sent. As in Rosbag2 we want to support the following features:
+
+* Save data to rosbag on trigger
+  * Enable triggering buffer save to disk with service call
+
+* Pause and resume buffering process
+  * Pausing refers to stopping buffering new messages until resumed or write is triggered.
+  * Resuming continues the process of buffering after pause. Resuming would clear all buffers.
+
+* Node and buffer configuration
+
+  Enable user to specify `cli` arguments which could be shared with `ros2 bag record` command. User should be able to configure the following arguments:
+
+    | parameter name | description   | behavior  |
+    | -------------- | ------------- | --------- |
+    | buffer size    | maximum memory per topic | <ul><li> defined by new `--max-buffer-size` argument, default value is 0 [bytes]  </li><li>if not set/set to non-positive value only duration parameter will be relevant (*note: either buffer size or duration must be specified*) </li></ul> |
+    | duration       | maximum time difference between newest and oldest message | <ul><li> defined by new `--max-buffer-duration` argument, default value is 10.0 [s] </li><li> if not set/set to non-positive value only buffer size parameter will be relevant (*note: either buffer size or duration must be specified*) </li></ul> |
```

Comment:
Same as above comment - I think we have a feature request for a maximum duration of the write cache, this should probably be the same parameter, to minimize the number of parameters.